autobuild/autobuild.8
.TH autobuild 8 "The Trusted Domain Project"
.SH NAME
.B autobuild
\- multiple configuration build tool
.SH SYNOPSIS
.B autobuild
[\-c] [\-n] [\-t] [\-v] descr-file
.SH DESCRIPTION
.B autobuild
configures, builds, and tests a source code package whose configuration is
provided by GNU's
.I autoconf
mechanism.
.I descr-file
contains a set of parameters to be passed to
.I configure
in various combinations, followed by
.I "make"
and
.I "make distcheck",
to confirm that the current source code tree has not had any build or runtime
regressions checked into it.

If any of the three steps fails, the output of the command is dumped
to standard output and execution terminates.
.SH OPTIONS
.TP
.I \-c
Print the configuration combinations, but do not execute them.
.TP
.I \-n
Parse configuration file and exit without doing anything else.
.TP
.I \-t
Print timestamps to stdout as each step is executed.
.TP
.I \-v
Prints configuration combinations as their commands are executed.
.SH DESCRIPTION FILE
The description file contains a JSON object.  It is expected to contain
optional keys, which refer to arrays, as follows:
.TP
.I always
Contains an array of strings each of which is an argument that should be passed
to
.I configure
for every test configuration.
.TP
.I combine
Contains an array of objects each of which is to be tested in various
combinations with all of the others in the array.  This means every entry
added to this array doubles the number of test configurations that will
be tried (i.e., the presence and absence cases).  An object in this array
must either be a string, or an array of strings; in the latter case, 
the set of strings is applied as a group whenever the array is selected for
testing.
.TP
.I each
Contains an array of objects each of which is to be tested in a mutually
exclusive way (that is, in sequence, but never together).  Exactly one 
will always be chosen and included.  To cause an iteration that chooses
none of the objects, include an empty object.
.SH EXAMPLE
{ "always": [ "--foo", "--bar" ], "combine": [ "--a", "--b", [ "--c", "--d"]] }
.PP
This will run eight tests; all of them will include "--foo" and "--bar",
and they will include all of the various possible combinationks of "--a", "--b",
and the pair "--c" and "--d".
.SH VERSION
This man page covers version 1.1.0 of
.I autobuild.
.SH COPYRIGHT
Copyright (c) 2012, 2013, The Trusted Domain Project.  All rights reserved.
.SH SEE ALSO
.B make(1)


RELEASE_NOTES
			OPENDKIM RELEASE NOTES

This listing shows the versions of the OpenDKIM package, the date of
release, and a summary of the changes in that release.

2.8.1		2013/0?/??
	Fix bug #SF3607071: Report the reason why a key file is determined
		to be unsafe.  Problem noted by Doug Barton.
	Fix bug #SF3607072: When checking for key file safety, take any
		"-u" value provided on the command line into account.
		Problem noted by Doug Barton.
	BUILD: Fix build for versions of libdb between 3.1 and 4.6.
		Problem noted by John Wood.

2.8.0		2013/02/25
	Feature request #SF2964383: Add DKIM_LIBFLAGS_STRICTRESIGN, which
		inhibits signing of a handle tagged for resigning when the
		attached verifying handle had no valid signatures in it.
	Feature request #SF3155117: Do a more thorough check for writeable
		key files, checking more of the filesystem permission tree.
	Feature request #SF3530734: Add "LDAPDisableCache", which suppresses
		the creation of a local cache in front of LDAP queries.
		Requested by Quanah Gibson-Mount.
	Feature request #SF3547359: If compiled with libcurl, add "SMTPURI"
		configuration option that allows direct SMTP transmission
		failure reports.  Requested by Andreas Schulze.
	Feature request #SF3578197: Allow per-message override of the list of
		header fields to be signed.  Requested by Alec Peterson.
	Feature request #SF3590860: Combine collected reputation values into
		an overall allowed rate under _FFR_REPRRD, as is done for the
		other reputation code.
	Feature request #SF3598991: Add odkim.signfor() function to the Lua
		setup script.  Requested by Marcin Owsiany.
	Feature request #SF3599409: Modify dkimf_checkip() to try surrounding
		the IP address part of every query with square brackets, which
		is a common way to do IP address literals in email contexts.
		Requested by Quanah Gibson-Mount.
	Fix bug #SF3531477: Add (hopefully temporary) configuration option
		"DisableCryptoInit" so that opendkim's initialization of the
		crypto library doesn't conflict with the same work done by
		other libraries.  Reported by Quanah Gibson-Mount.
	Fix bug #SF3599901: Rename "InsecureKey" to "UnprotectedKey" and
		"InsecurePolicy" to "UnprotectedPolicy", as the term "insecure"
		in reference to a key is sometimes interpreted to mean "not
		enough random bits" rather than as a keyword describing the
		presence or absence of DNSSEC protection.  What's logged in
		Authentication-Results header fields has been similarly
		modified.  Suggested by Scott Kitterman.
	Fix bug #SF3604525: Don't divide by zero when the query cache hasn't
		been used.  Reported by Denis Klimov.
	Protect against handling of signatures with empty domains, which could
		cause a NULL dereference and a crash.  Problem noted by
		Motohiro Ishiyama and John Wood.
	Do ATPS checks when enabled even if ADSP is disabled.
 	Don't fail to start on empty or null configuration files.  Problem
 		noted by Steve Jenkins.
	Patch #SF3593422: Update for MDB 0.9.5 support.  Patch from
		Quanah Gibson-Mount.
	LIBOPENDKIM: Fix header canonicalization when DKIM_LIBFLAG_FIXCRLF is
		used in combination with dkim_chunk().  Problem noted by
		Dave Kelly and Heather Lord.
	LIBOPENDKIM: Enable dkim_getcachestats() and the underlying function
		to extract the current number of keys in the cache, and also
		provide a counter reset mechanism.
	BUILD: Feature request #SF3547151: Check for Lua package name variants
		in use on Debian.  Requested by Scott Kitterman.
	BUILD: Feature request #SF3599902: Change OpenSSL existence test
		to help with Debian packaging.  Requested by Scott Kitterman.
	BUILD: Add "--with-test-socket" to force all of the filter unit tests
		to use a specific socket.  Based on a bug report from
		Scott Kitterman.
	BUILD: Add checks for strlcat()/strlcpy() in libbsd.  Patch from
		Scott Kitterman.
	CONTRIB: Fix bug #SF3575666: Pass pid file path to killproc.
		Suggested by Christophe Wolfhugel.
	CONTRIB: Add systemd directory.  Contributed by Steve Jenkins.
	CONTRIB: Split out initial key generation function from
		contrib/init/redhat/opendkim.  Contributed by Steve Jenkins.
	MILTERTEST: Don't crash in mt_connect() if the socketspec doesn't
		contain a colon.
	MILTERTEST: When connect() fails for an AF_INET socket, it apparently
		leaves the socket unusable.  Discard the socket when that
		happens and get a new one.
	MILTERTEST: Add a way to extend the mt.connect() retry interval via
		environment variables so a large test suite can be easily
		extended on slow systems.  Problem noted by Scott Kitterman.
	TOOLS: Register DNS functions before calling dkim_dns_init() in
		opendkim-testkey.  Problem noted by Jeff Anton.
	TOOLS: Add "-K" (keep temporary files) flag for opendkim-testmsg.

2.7.4		2013/01/06
	Fix bug #SF3596147: Allow arbitrarily long configuration file lines.
		Based on a patch from Daniel Black.
	Fix bug #SF3596229: Fix logging of signature errors, which logged
		the domain name twice instead of the domain name and the
		selector.  Patch from Daniel Black.
	Safely handle incoming Authentication-Results fields with large number
		of tokens.  Problem noted by Motohiro Ishiyama and John Wood.
	Avoid mixing up status codes when processing conf_logresults, leading
		to confusing log entries.  Problem reported by John Wood.
	Fix ADSP logging.  Problem noted by Claus Assmann.

2.7.3		2012/11/29
	Log DB error string in dkimf_add_signrequest(), and fix a DSN handling
		error in dkimf_db_strerror().  Problem noted by
		Simone Caruso.
	LIBOPENDKIM: Ignore entries in the oversign header field name list
		that are empty, and an oversign header field name list that
		is present but empty.  Problem noted by Alec Peterson.
	LIBOPENDKIM: Allow header field lists to be empty, flushing any that
		were previously defined.  Problem noted by Alec Peterson.
	BUILD: Improve tests for including <strl.h>.  Based on a patch from
		Eray Aslan.
	REPUTATION: Use lowercase for keywords in REPUTE query generation
		and handling.

autobuild/autobuild.c
/*
**  Copyright (c) 2012, 2013, The Trusted Domain Project.
**  	All rights reserved.
*/

#include <sys/param.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <sysexits.h>
#include <time.h>
#include <string.h>
#include <stdio.h>
#include <math.h>
#include <errno.h>

#include <jansson.h>

#define	JSON_ALWAYS	"always"
#define	JSON_COMBINE	"combine"
#define	JSON_EACH	"each"

#define	AUTORECONF	"autoreconf"
#define	CONFIGURE	"./configure"
#define	MAKE		"make"
#define	DISTCHECK	"distcheck"
#define	CLEAN		"clean"

#define	BUFRSZ		2048
#define	TEMPLATE	"/tmp/abXXXXXX"

#ifndef MIN
# define MIN(a,b)	((a) < (b) ? (a) : (b))
#endif /* ! MIN */
#ifndef MAX
# define MAX(a,b)	((a) > (b) ? (a) : (b))
#endif /* ! MAX */

char *progname;

/*
**  DUMPARGS -- dump argument vector
**
**  Parameters:
**  	where -- output stream
**  	args -- argument vector
**
**  Return value:
**  	None.
*/

void
dumpargs(FILE *where, const char **args)
{
	int n;

	for (n = 0; args[n] != NULL; n++)
		fprintf(where, "%s%s\n", n == 0 ? "" : "\t", args[n]);
}

/*
**  USAGE -- print usage message
**
**  Parameters:
**  	None.
**
**  Return value:
**  	EX_USAGE
*/

int
usage(void)
{
	fprintf(stderr, "%s: usage: %s [options] descr-file\n"
	                "\t-c\tshow combinations and exit\n"
	                "\t-n\tparse descr-file and exit\n"
	                "\t-t\tshow timestamps\n"
	                "\t-v\tverbose mode\n", progname, progname);

	return EX_USAGE;
}

/*
**  MAIN -- program mainline
**
**  Parameters:
**  	argc, argv -- the usual
**
**  Return value:
**  	Exit status.
*/

int
main(int argc, char **argv)
{
	int status;
	int m;
	int n;
	int fd;
	int showcombos = 0;
	int verbose = 0;
	int timestamps = 0;
	int confonly = 0;
	pid_t child;
	size_t combos;
	size_t asz;
	size_t c;
	size_t d;
	size_t bits;
	size_t nopts;
	size_t neach = 1;
	size_t nargs;
	size_t maxopts;
	size_t meach = 0;
	size_t xeach;
	json_t *j;
	json_t *node;
	json_t *always = NULL;
	json_t *combine = NULL;
	json_t *each = NULL;
	json_error_t err;
	void *iter;
	char *p;
	char *key;
	char *descr;
	const char **args;
	char buf[BUFRSZ];
	char fn[MAXPATHLEN + 1];

	progname = (p = strrchr(argv[0], '/')) == NULL ? argv[0] : p + 1;

	while ((c = getopt(argc, argv, "cntv")) != -1)
	{
		switch (c)
		{
		  case 'c':
			showcombos++;
			break;

		  case 'n':
			confonly++;
			break;

		  case 't':
			timestamps++;
			break;

		  case 'v':
			verbose++;
			break;

		  default:
			return usage();
		}
	}

	if (optind >= argc)
		return usage();

	descr = argv[optind];
	j = json_load_file(descr, 0, &err);
	if (j == NULL)
	{
		fprintf(stderr, "%s: JSON input error from %s:\n"
		                "\ttext = \"%s\"\n"
		                "\tsource = \"%s\"\n"
		                "\tline = %d\n"
		                "\tcolumn = %d\n"
		                "\toffset = %ld\n",
		        progname, descr, err.text, err.source, err.line,
		        err.column, err.position);
		return EX_DATAERR;
	}

	/* the top has to be an object */
	if (!json_is_object(j))
	{
		fprintf(stderr, "%s: %s: root object is not an object\n",
		        progname, descr);
		json_decref(j);
		return EX_DATAERR;
	}

	/* ensure there are no top-level objects we don't expect */
	for (iter = json_object_iter(j);
	     iter != NULL;
	     iter = json_object_iter_next(j, iter))
	{
		key = (char *) json_object_iter_key(iter);
		node = json_object_iter_value(iter);

		if (strcasecmp(key, JSON_ALWAYS) != 0 &&
		    strcasecmp(key, JSON_EACH) != 0 &&
		    strcasecmp(key, JSON_COMBINE) != 0)
		{
			fprintf(stderr,
			        "%s: %s: unexpected root object \"%s\"\n",
			        progname, descr, key);
			json_decref(j);
			return EX_DATAERR;
		}
		else if (!json_is_array(node))
		{
			fprintf(stderr,
			        "%s: %s: root object \"%s\" is not an array\n",
			        progname, descr, key);
			json_decref(j);
			return EX_DATAERR;
		}
		else if (strcasecmp(key, JSON_ALWAYS) == 0)
		{
			always = node;
		}
		else if (strcasecmp(key, JSON_EACH) == 0)
		{
			each = node;
		}
		else
		{
			combine = node;
		}
	}

	if (confonly > 0)
	{
		json_decref(j);
		return EX_OK;
	}

	nargs = 2;
	if (always != NULL)
		nargs += json_array_size(always);
	if (combine == NULL)
		nopts = 0;
	else
		nopts = json_array_size(combine);
	if (each == NULL)
	{
		neach = 1;
<<<<<<< HEAD
=======
		meach = 0;
>>>>>>> 2409b120b5b863479827ed16aade83763372af11
	}
	else
	{
		neach = json_array_size(each);
		for (n = 0; n < neach; n++)
		{
			node = json_array_get(each, n);
			if (json_is_array(node))
				meach = MAX(json_array_size(node), meach);
		}
	}

	for (n = 0; n < nopts; n++)
	{
		node = json_array_get(combine, n);

		switch (json_typeof(node))
		{
		  case JSON_ARRAY:
			for (m = 0; m < json_array_size(node); m++)
			{
				if (!json_is_string(json_array_get(node, m)))
				{
					fprintf(stderr,
					        "%s: %s: combine object at index %d is an array containing non-strings\n",
					        progname, descr, n);
					return EX_DATAERR;
				}

				nargs++;
			}

			break;

		  case JSON_STRING:
			nargs++;
			break;

		  default:
			fprintf(stderr,
			        "%s: %s: combine object at index %d is not a string or array\n",
			        progname, descr, n);
			return EX_DATAERR;
		}
	}

	asz = sizeof(char *) * (nargs + nopts + meach + 2);
	args = (const char **) malloc(asz);
	if (args == NULL)
	{
		fprintf(stderr, "%s: malloc(): %s\n", progname,
		        strerror(errno));
		return EX_OSERR;
	}

	combos = (size_t) pow(2, nopts) * neach;

	if (verbose > 1)
	{
		fprintf(stdout, "%s: max %d arguments, %d combinations\n",
		        progname, nargs + nopts, combos);
	}

	/* autoreconf */
	args[0] = AUTORECONF;
	args[1] = "-v";
	args[2] = "-i";
	args[3] = NULL;

	if (!showcombos)
	{
		strncpy(fn, TEMPLATE, sizeof fn);
		fd = mkstemp(fn);
		if (fd < 0)
		{
			fprintf(stderr, "%s: mkstemp(): %s\n", progname,
		        	strerror(errno));
			return EX_OSERR;
		}
		(void) unlink(fn);

		child = fork();
		switch (child)
		{
		  case -1:
			fprintf(stderr, "%s: fork(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  case 0:
			(void) dup2(fd, 1);
			(void) dup2(fd, 2);
			(void) execvp(args[0], (char * const *) args);
			fprintf(stderr, "%s: execvp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  default:
			n = wait(&status);
			if (n == -1)
			{
				fprintf(stderr, "%s: wait(): %s\n", progname,
				        strerror(errno));
				return EX_OSERR;
			}
			else if (WIFSIGNALED(status) ||
			         WEXITSTATUS(status) != 0)
			{
				if (WIFSIGNALED(status))
				{
					fprintf(stderr,
					        "%s: clean died with signal %d\n",
					        progname, WTERMSIG(status));
				}
				else
				{
					fprintf(stderr,
					        "%s: clean exited with status %d\n",
					        progname, WEXITSTATUS(status));
				}

				(void) lseek(fd, 0, SEEK_SET);

				dumpargs(stdout, args);

				for (;;)
				{
					n = read(fd, buf, sizeof buf);
					(void) fwrite(buf, 1, n, stdout);
					if (n < sizeof buf)
						break;
				}

				close(fd);
				free(args);
				json_decref(j);
				return 1;
			}

			break;
		}

		/* clean up */
		close(fd);
	}

	for (c = 0; c < combos; c++)
	{
		xeach = c % neach;

		memset(args, '\0', asz);
		bits = c;

		args[0] = CONFIGURE;
		n = 1;

		/* add the "always" arguments */
		for (d = 0; d < json_array_size(always); d++)
		{
			node = json_array_get(always, d);
			args[n++] = json_string_value(node);
		}

		/* select the "each" argument */
		if (each != NULL)
		{
			node = json_array_get(each, xeach);
			if (json_is_string(node))
			{
				args[n++] = json_string_value(node);
			}
			else
			{
				json_t *sub;

				for (m = 0;
				     m < json_array_size(node);
				     m++)
				{
					sub = json_array_get(node, m);
					args[n++] = json_string_value(sub);
				}
			}
		}

		/* add the new combination of options */
		for (d = 0; d < nopts; d++)
		{
			if ((c / neach) & (1 << d))
			{
				node = json_array_get(combine, d);

				if (json_is_string(node))
				{
					args[n++] = json_string_value(node);
				}
				else
				{
					json_t *sub;

					for (m = 0;
					     m < json_array_size(node);
					     m++)
					{
						sub = json_array_get(node, m);
						args[n++] = json_string_value(sub);
					}
				}
			}
		}

		if (showcombos)
		{
			dumpargs(stdout, args);
			continue;
		}

		if (verbose)
		{
			if (timestamps)
			{
				time_t now;

				(void) time(&now);
				fprintf(stdout, "%s: %s", progname,
				        ctime(&now));
			}
			
			fprintf(stdout, "%s: ", progname);
			dumpargs(stdout, args);
		}

		/* ./configure */
		strncpy(fn, TEMPLATE, sizeof fn);
		fd = mkstemp(fn);
		if (fd < 0)
		{
			fprintf(stderr, "%s: mkstemp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;
		}
		(void) unlink(fn);

		child = fork();
		switch (child)
		{
		  case -1:
			fprintf(stderr, "%s: fork(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  case 0:
			(void) dup2(fd, 1);
			(void) dup2(fd, 2);
			(void) execvp(args[0], (char * const *) args);
			fprintf(stderr, "%s: execvp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  default:
			n = wait(&status);
			if (n == -1)
			{
				fprintf(stderr, "%s: wait(): %s\n", progname,
				        strerror(errno));
				return EX_OSERR;
			}
			else if (WIFSIGNALED(status) ||
			         WEXITSTATUS(status) != 0)
			{
				if (WIFSIGNALED(status))
				{
					fprintf(stderr,
					        "%s: configure died with signal %d\n",
					        progname, WTERMSIG(status));
				}
				else
				{
					fprintf(stderr,
					        "%s: configure exited with status %d\n",
					        progname, WEXITSTATUS(status));
				}

				(void) lseek(fd, 0, SEEK_SET);

				dumpargs(stdout, args);

				for (;;)
				{
					n = read(fd, buf, sizeof buf);
					(void) fwrite(buf, 1, n, stdout);
					if (n < sizeof buf)
						break;
				}

				close(fd);
				free(args);
				json_decref(j);
				return 1;
			}

			break;
		}

		close(fd);

		/* make */
		args[0] = MAKE;
		args[1] = NULL;

		strncpy(fn, TEMPLATE, sizeof fn);
		fd = mkstemp(fn);
		if (fd < 0)
		{
			fprintf(stderr, "%s: mkstemp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;
		}
		(void) unlink(fn);

		child = fork();
		switch (child)
		{
		  case -1:
			fprintf(stderr, "%s: fork(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  case 0:
			(void) dup2(fd, 1);
			(void) dup2(fd, 2);
			(void) execvp(args[0], (char * const *) args);
			fprintf(stderr, "%s: execvp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  default:
			n = wait(&status);
			if (n == -1)
			{
				fprintf(stderr, "%s: wait(): %s\n", progname,
				        strerror(errno));
				return EX_OSERR;
			}
			else if (WIFSIGNALED(status) ||
			         WEXITSTATUS(status) != 0)
			{
				if (WIFSIGNALED(status))
				{
					fprintf(stderr,
					        "%s: make died with signal %d\n",
					        progname, WTERMSIG(status));
				}
				else
				{
					fprintf(stderr,
					        "%s: make exited with status %d\n",
					        progname, WEXITSTATUS(status));
				}

				(void) lseek(fd, 0, SEEK_SET);

				dumpargs(stdout, args);

				for (;;)
				{
					n = read(fd, buf, sizeof buf);
					(void) fwrite(buf, 1, n, stdout);
					if (n < sizeof buf)
						break;
				}

				close(fd);
				free(args);
				json_decref(j);
				return 1;
			}

			break;
		}

		close(fd);

		/* make distcheck */
		args[0] = MAKE;
		args[1] = DISTCHECK;
		args[2] = NULL;

		strncpy(fn, TEMPLATE, sizeof fn);
		fd = mkstemp(fn);
		if (fd < 0)
		{
			fprintf(stderr, "%s: mkstemp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;
		}
		(void) unlink(fn);

		child = fork();
		switch (child)
		{
		  case -1:
			fprintf(stderr, "%s: fork(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  case 0:
			(void) dup2(fd, 1);
			(void) dup2(fd, 2);
			(void) execvp(args[0], (char * const *) args);
			fprintf(stderr, "%s: execvp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  default:
			n = wait(&status);
			if (n == -1)
			{
				fprintf(stderr, "%s: wait(): %s\n", progname,
				        strerror(errno));
				return EX_OSERR;
			}
			else if (WIFSIGNALED(status) ||
			         WEXITSTATUS(status) != 0)
			{
				if (WIFSIGNALED(status))
				{
					fprintf(stderr,
					        "%s: distcheck died with signal %d\n",
					        progname, WTERMSIG(status));
				}
				else
				{
					fprintf(stderr,
					        "%s: distcheck exited with status %d\n",
					        progname, WEXITSTATUS(status));
				}

				(void) lseek(fd, 0, SEEK_SET);

				dumpargs(stdout, args);

				for (;;)
				{
					n = read(fd, buf, sizeof buf);
					(void) fwrite(buf, 1, n, stdout);
					if (n < sizeof buf)
						break;
				}

				close(fd);
				free(args);
				json_decref(j);
				return 1;
			}

			break;
		}

		/* clean up */
		close(fd);

		/* make clean */
		args[0] = MAKE;
		args[1] = CLEAN;
		args[2] = NULL;

		strncpy(fn, TEMPLATE, sizeof fn);
		fd = mkstemp(fn);
		if (fd < 0)
		{
			fprintf(stderr, "%s: mkstemp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;
		}
		(void) unlink(fn);

		child = fork();
		switch (child)
		{
		  case -1:
			fprintf(stderr, "%s: fork(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  case 0:
			(void) dup2(fd, 1);
			(void) dup2(fd, 2);
			(void) execvp(args[0], (char * const *) args);
			fprintf(stderr, "%s: execvp(): %s\n", progname,
			        strerror(errno));
			return EX_OSERR;

		  default:
			n = wait(&status);
			if (n == -1)
			{
				fprintf(stderr, "%s: wait(): %s\n", progname,
				        strerror(errno));
				return EX_OSERR;
			}
			else if (WIFSIGNALED(status) ||
			         WEXITSTATUS(status) != 0)
			{
				if (WIFSIGNALED(status))
				{
					fprintf(stderr,
					        "%s: clean died with signal %d\n",
					        progname, WTERMSIG(status));
				}
				else
				{
					fprintf(stderr,
					        "%s: clean exited with status %d\n",
					        progname, WEXITSTATUS(status));
				}

				(void) lseek(fd, 0, SEEK_SET);

				dumpargs(stdout, args);

				for (;;)
				{
					n = read(fd, buf, sizeof buf);
					(void) fwrite(buf, 1, n, stdout);
					if (n < sizeof buf)
						break;
				}

				close(fd);
				free(args);
				json_decref(j);
				return 1;
			}

			break;
		}

		/* clean up */
		close(fd);
	}

	return 0;
}
